File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/converter/DefaultValueConverter.java
package org.skellig.teststep.processing.converter;

import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.valueextractor.TestStepValueExtractor;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;

public class DefaultValueConverter implements TestStepValueConverter {

    private List<TestStepValueConverter> valueConverters;

    protected DefaultValueConverter(List<TestStepValueConverter> valueConverters) {
        this.valueConverters = valueConverters;
    }

    @Override
    public Object convert(String value) {
        Object result = value;
        for (TestStepValueConverter valueConverter : valueConverters) {
            if (result != null && result.getClass().equals(String.class)) {
                result = valueConverter.convert((String) result);
            }
        }
        return result;
    }

    public static class Builder {

        private List<TestStepValueConverter> valueConverters;
        private TestScenarioState testScenarioState;
        private TestStepValueExtractor testStepValueExtractor;
        private Function<String, String> getPropertyFunction;
        private ClassLoader classLoader;

        public Builder() {
            valueConverters = new ArrayList<>();
        }

        public Builder withTestScenarioState(TestScenarioState testScenarioState) {
            this.testScenarioState = testScenarioState;
            return this;
        }

        public Builder withGetPropertyFunction(Function<String, String> getPropertyFunction) {
            this.getPropertyFunction = getPropertyFunction;
            return this;
        }

        public Builder withClassLoader(ClassLoader classLoader) {
            this.classLoader = classLoader;
            return this;
        }

        public Builder withTestStepValueExtractor(TestStepValueExtractor testStepValueExtractor) {
            this.testStepValueExtractor = testStepValueExtractor;
            return this;
        }

        public Builder withValueConverter(TestStepValueConverter valueConverter) {
            this.valueConverters.add(valueConverter);
            return this;
        }

        public TestStepValueConverter build() {
            Objects.requireNonNull(testScenarioState, "Test Scenario State must be provided");

            valueConverters.add(0, new PropertyValueConverter(valueConverters, getPropertyFunction));
            withValueConverter(new TestStepStateValueConverter(testScenarioState, testStepValueExtractor));
            if (classLoader != null) {
                withValueConverter(new FileValueConverter(classLoader));
            }
            withValueConverter(new NumberValueConverter());
            withValueConverter(new IncrementValueConverter());
            withValueConverter(new DateTimeValueConverter());

            return new DefaultValueConverter(valueConverters);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/converter/DefaultValueConverter.kt
package org.skellig.teststep.processing.converter

import org.skellig.teststep.processing.experiment.FunctionValueProcessor
import org.skellig.teststep.processing.experiment.ValueExtractor
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.valueextractor.TestStepValueExtractor
import java.util.*

class DefaultValueConverter private constructor(private val functions: Map<String, FunctionValueProcessor>) : FunctionValueProcessor {

    override fun execute(name: String, args: Array<Any?>): Any? = functions[name]?.execute(name, args)

    override fun getFunctionName(): String = ""

    class Builder {

        private val valueConverters = mutableListOf<TestStepValueConverter>()
        private val functions = mutableMapOf<String, FunctionValueProcessor>()
        private var testScenarioState: TestScenarioState? = null
        private var valueExtractor: ValueExtractor? = null
        private var getPropertyFunction: ((String) -> Any?)? = null
        private var classLoader: ClassLoader? = null
        private var classPaths: Collection<String>? = null

        fun withTestScenarioState(testScenarioState: TestScenarioState?) =
            apply { this.testScenarioState = testScenarioState }

        fun withGetPropertyFunction(getPropertyFunction: ((String) -> Any?)?) =
            apply { this.getPropertyFunction = getPropertyFunction }

        fun withClassLoader(classLoader: ClassLoader?) = apply { this.classLoader = classLoader }

        fun withClassPaths(classPaths: Collection<String>) = apply { this.classPaths = classPaths }

        fun withTestStepValueExtractor(valueExtractor: ValueExtractor?) =
            apply { this.valueExtractor = valueExtractor }

        fun withValueConverter(valueConverter: TestStepValueConverter) = apply { valueConverters.add(valueConverter) }

        fun withFunctionProcessor(functionProcessor: FunctionValueProcessor) = apply { functions[functionProcessor.getFunctionName()] = functionProcessor }

        fun build(): FunctionValueProcessor {
            Objects.requireNonNull(testScenarioState, "Test Scenario State must be provided")

            val defaultValueConverter = DefaultValueConverter(functions)

            withFunctionProcessor(TestDataFromIfStatementConverter())
            withFunctionProcessor(TestStepStateValueConverter(testScenarioState!!))
//            withFunctionProcessor(FindFromStateValueConverter(testScenarioState!!, testStepValueExtractor))
            classLoader?.let {
                withFunctionProcessor(FileValueConverter(it))
            }

            withFunctionProcessor(RandomValueConverter())
            withFunctionProcessor(IncrementValueConverter())
            withFunctionProcessor(CurrentDateTimeValueConverter())
            withFunctionProcessor(ToDateTimeValueConverter())
            withFunctionProcessor(ListOfValueConverter())
            withFunctionProcessor(TestDataToJsonConverter())
            classLoader?.let {
                val testDataFromCsvConverter = TestDataFromCsvConverter(it)
               withFunctionProcessor(testDataFromCsvConverter)
               withFunctionProcessor(TestDataFromCsvConverter(it))
               withFunctionProcessor(TestDataFromFTLConverter(it, testDataFromCsvConverter))
               withFunctionProcessor(CustomFunctionValueConverter(classPaths, it))
            }

            return defaultValueConverter
        }
    }
}